packages/actor-query-operation-minus/lib/ActorQueryOperationMinus.ts

```diff
@@ -0,0 +1,64 @@
+import {AbstractBindingHash, IActorInitRdfDereferencePagedArgs} from "@comunica/actor-abstract-bindings-hash";
+import {
+    ActorQueryOperation, Bindings, BindingsStream,
+    IActorQueryOperationOutputBindings,
+} from "@comunica/bus-query-operation";
+import {ActionContext} from "@comunica/core";
+import * as RDF from "rdf-js";
+import {Algebra} from "sparqlalgebrajs";
+
+/**
+ * A comunica Minus Query Operation Actor.
+ */
+export class ActorQueryOperationMinus extends AbstractBindingHash<Algebra.Minus> {
+
+  private hashes: {[id: string]: boolean} = {};
+  private commons: string[];
+
+  constructor(args: IActorInitRdfDereferencePagedArgs) {
+    super(args, 'minus');
+  }
+
+  public newHashFilter(hashAlgorithm: string, digestAlgorithm: string)
+        : (bindings: Bindings) => boolean {
+    return (bindings: Bindings) => {
+      const hash: string = ActorQueryOperationMinus.hash(hashAlgorithm, digestAlgorithm,
+          bindings.filter((v: RDF.Term, k: string) => -1 !== this.commons.indexOf(k)));
+
+      return !(hash in this.hashes);
+    };
+  }
+
+  public async runOperation(pattern: Algebra.Minus, context: ActionContext)
+      : Promise<IActorQueryOperationOutputBindings> {
+
+    const buffer = ActorQueryOperation.getSafeBindings(
+          await this.mediatorQueryOperation.mediate({ operation: pattern.right, context }));
+    const output = ActorQueryOperation.getSafeBindings(
+          await this.mediatorQueryOperation.mediate({ operation: pattern.left, context }));
+
+    this.commons = buffer.variables.filter((n: string) => -1 !== output.variables.indexOf(n));
+    if (this.commons.length > 0) {
+      let bindingsStream: BindingsStream = null;
+      const prom = new Promise((resolve) => {
```

Comment:
As far as I understand it, you use this promise to make the execution block until the left stream is finished.

I suggest you look into the usages of `PromiseProxyIterator` in this project, and solve this in a similar way.
The reason for this is that we want to reach the `return` statement of this method as soon as possible. The streams however, are allowed to block at any time.

Reply:
So to specificy, since there was some confusion about this, the current problem is that the return statement of this function can not be reached before the entire right side (buffer) is drained. This would cause everything else to be stuck, waiting on this actor. The problem is the `await prom;`. This should not happen before the output object is returned. This means the bindingsStream should actually return a promise calculating all this stuff (and since its type is an iterator and not a promise, you should use a `PromiseProxyIterator` to encapsulate).